### PR TITLE
ArC: add synthetic interceptors API

### PR DESCRIPTION
**Pull request by mkouba**
- [x] add tests for other interception types

### CODE REVIEW

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InterceptorCreator.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    @FunctionalInterface
+    public interface InterceptFunction {
+
+        /**
+         *
+         * @param invocationContext
+         * @return the return value
+         * @throws Exception
+         */
+        Object intercept(InvocationContext invocationContext) throws Exception;
```
Comment (manovotn) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InterceptorCreator.java:26]: I'd use `ArcInvocationContext` as a parameter here just to give users access to additional functionality we have there.

Comment (Ladicek) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InterceptorCreator.java:26]: Good point. I think I'm +1 to that.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/InterceptorGenerator.java [hunk 2]
```diff
@@ -235,8 +323,7 @@ private void addIntercepts(InterceptorInfo interceptor, InterceptionType interce
                     .readStaticField(FieldDescriptor.of(InterceptionType.class.getName(), interceptionType.name(),
                             InterceptionType.class.getName()));
             BranchResult result = intercepts
-                    .ifNonZero(intercepts.invokeVirtualMethod(MethodDescriptors.OBJECT_EQUALS, enumValue,
-                            intercepts.getMethodParam(0)));
+                    .ifTrue(Gizmo.equals(intercepts, enumValue, intercepts.getMethodParam(0)));
             result.trueBranch().returnValue(result.trueBranch().load(true));
```
[the remark this answers is not in the document]
Comment (mkouba) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/InterceptorGenerator.java]: Done.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InterceptorCreator.java (hunk 1 above)
Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InterceptorCreator.java]: Done.

**Comment by mkouba** [comment]
> LGTM, but tests for other interceptor types should indeed be added.

And done ;-).

**Review by Ladicek — APPROVED** [review]

**Comment by Ladicek** [comment]
Nice!

**Review by manovotn — APPROVED** [review]
LGTM!

**Pull request merged**